app/controllers/save_pages_controller.rb

```diff
@@ -0,0 +1,35 @@
+class SavePagesController < ApplicationController
+  include AccountConcern
+
+  before_action :set_no_cache_headers
+  before_action { head :not_found unless feature_flag_enabled? }
+  before_action :authenticate
+
+  def create
+    GdsApi.account_api.save_page(page_path: params[:page_path], govuk_account_session: account_session_header)
+    redirect_to params[:page_path] + "?personalisation=page_saved"
+  rescue GdsApi::HTTPUnauthorized
+    @account_session_header = nil
+    authenticate
+  rescue GdsApi::HTTPUnprocessableEntity
+    redirect_to params[:page_path] + "?personalisation=page_not_saved"
+  end
+
+private
+
+  def authenticate
+    if feature_flag_enabled?
+      unless logged_in?
+        redirect_with_ga new_govuk_session_path(
+          redirect_path: "/save-a-page?page_path=#{params[:page_path]}",
+          state_id: params[:state_id],
```

Comment:
That `state_id` is only used to make the Transition Checker kind-of-not-a-JWT-but-also-is-definitely-a-JWT journey work, it's only needed in the `SessionsController`, not here.

Reply:
oh great! I wasn't sure so I thought i'd include it.
Will clear that out now.

Reply:
Also, we could eliminate the redirect via `/sign-in` by using `GdsApi.account_api.get_sign_in_url` directly here:

```ruby
  def authenticate
    return if !feature_flag_enabled? || logged_in?

    redirect_with_ga GdsApi.account_api.get_sign_in_url(redirect_path: "...")
  end
```

Removing one redirect from the chain is only a slight improvement, but feels like a nice one.

Reply:
Cool, saves us one more hop - I figured we might need to remember to change this if we ever adjusted /sign-in, but that seems like a manageable thing.

Done!